### DISCUSSION TITLE
Template project for automatically building, packaging and publishing a Djinni library

### DISCUSSION THREAD

**Discussion by jothepro** [cross-language-cpp/djinni-generator#88] [Show and tell]
I'd like to show you what I've built in the past months: https://github.com/jothepro/djinni-library-template

It's a template for a Djinni library that can be used on Android, iOS/macOS and Windows without having to mess with all the nasty platform and build-tool specific details.

It comes with a set of utilities that automate the building, packaging and publishing of a Djinni library. It can create bundles for ...
- macOS, iOS (XCFramework)
- Android (AAR)
- Windows .NET 5.0 (NuGet)
- plain C++ (Conan)

The resulting library can then be used in the platform-specific project like any normal library. No knowledge about Djinni and it's internals are required!

I've put a lot of research and time into putting all those pieces together and have started using it on a personal project successfully. 

I'd love to know what you think about my approach!
- Does this template cover your requirements? If not, what is it missing?
- Did you build something similar? If not, what's your approach to use/build Djinni interfaces?
- Do you miss any information in the docs?

**Comment by a4z** [comment]
Very good stuff, excellent know how compilation!

We have the general problem of dealing with a not trivial situation, cross compile for various systems, some of them mobile and special.
This does make every documentation difficult It is either not documented enough, or too huge and intimidating for beginners.

Maybe it could help to split the targets a little bit more

* for Android development -> Read here
* for iOS development -> Read here
* for Python development -> Read here 
* for C# development -> Read here

With a small sample project / goal people can work through, having something to execute at the end.
Your template looks like a good starting point for creating such intros.

I know you have it basically, but a sample project that tells the story, from the begin to the end (having something on a device), would for sure add additional value and lower entry barriers.
Maybe we can create something like that with the time.

**Comment by freitass** [comment]
What a great way to showcase Djinni's versatility! Like Harald said, it's great for someone to be able to see how to get the full power of Djinni from A to Z. (I've already got to know things I didn't know existed, like GitHub's NuGet repository!)

Something that draws my attention though is the number of "moving pieces": Git submodules, technologies (e.g. CMake and Python in the build infrastructure) -- even for a seasoned developer it can be a bit overwhelming. I'd consider, for instance, making something like https://github.com/jothepro/djinni-cmake available in the support-lib.

In any case, this is impressive work. Great job!
